### PR TITLE
`error-stack` add `serde` based serialization

### PR DESCRIPTION
**Pull request by indietyp**
## 🌟 What is the purpose of this PR?

Create a serde serialization implementation for `Report`.

## 🔗 Related links

- #792: Previous PR, due to internal changes closed in favor of this one

## 🔍 What does this change?

* non-derive based `serde::Serialize` implementation, which is gated behind a new feature flag

## 📜 Does this require a change to the docs?

Yes, while teased in the `CHANGELOG.md`, the `lib.rs` documentation and module level documentation need to be adjusted.

## 🐾 Next steps

* [x] Tests
* [ ] (Follow-Up PR): Hooks + Documentation

## 🛡 What tests cover this?

New snapshot tests that are located in `tests/test_ser.rs`

## 📹 Demo

```json5
[
  {
    "value": "context A",
    "attachments": [
      "printable C: 4",
    ],
    "sources": [
      {
        "value": "root error",
        "attachments": [
          "printable C: 1",
          "printable C: 3",
        ],
        "sources": [],
      },
      {
        "value": "root error",
        "attachments": [
          "printable C: 2",
          "printable C: 3",
        ],
        "sources": [],
      },
    ],
  },
]
```

### CODE REVIEW

**Comment by indietyp** [comment]
(Tests fail because the location for some snapshots changed, this will be fixed as soon as I am able to do so)

**Review by TimDiekmann — CHANGES_REQUESTED** [review]
@indietyp and I spoke about the exact representation offline and after a few back and forth agreed on a similar representation as we use internally. The contexts will be nested as this is the most intuitive format when a `Report` is serialized. This has the disadvantage, that long chains of errors will get bigger nesting, but we agreed that serialized reports are mainly used internally, and parsing those should be as straightforward as possible. To pretty print a report, the `Debug` implementation typically will be used.
The attachments will remain a list inside of each context.

File: packages/libs/error-stack/src/lib.rs [hunk 1]
```diff
@@ -477,6 +477,8 @@ pub mod fmt;
 mod fmt;
 #[cfg(feature = "std")]
 mod hook;
+#[cfg(feature = "serde")]
+mod ser;
```
Comment (TimDiekmann) [inline packages/libs/error-stack/src/lib.rs:481]: Nit: Do we want to rename the module to `serde`? I don't feel strongly, just want to flag.

File: packages/libs/error-stack/src/ser.rs [hunk 2]
```diff
@@ -0,0 +1,152 @@
+//! Implementation of general report serialization.
+//!
+//! The value can be of any type, currently only printable attachments and context are supported, in
+//! the near future any values will be supported through the use of hooks.
+//!
+//! The returned JSON returned is a list of all current sources with the following output:
```
Comment (TimDiekmann) [inline packages/libs/error-stack/src/ser.rs:6]: Serialization is not only JSON
```suggestion
//! The serialized [`Report`] is a list of all current sources with the following output:
```

File: packages/libs/error-stack/src/ser.rs [hunk 3]
```diff
@@ -0,0 +1,152 @@
+//! Implementation of general report serialization.
```
Comment (TimDiekmann) [inline packages/libs/error-stack/src/ser.rs:1]:
```suggestion
//! Implementation of general [`Report`] serialization.
```

File: packages/libs/error-stack/src/ser.rs [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+//!
+//! The value can be of any type, currently only printable attachments and context are supported, in
+//! the near future any values will be supported through the use of hooks.
+//!
+//! The returned JSON returned is a list of all current sources with the following output:
+//!
+//! ```json
+//! {
+//!     "context": "string value",
```
Comment (TimDiekmann) [inline packages/libs/error-stack/src/ser.rs:10]:
```suggestion
//!     "context": "context display output",
```

File: packages/libs/error-stack/src/ser.rs [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    where
+        S: Serializer,
+    {
+        let Self(frame) = self;
+
+        #[allow(clippy::match_same_arms)]
+        match frame.kind() {
+            FrameKind::Context(_) => {
+                // `SerializeContext` ensures that no context is ever serialized
+                unreachable!()
```
Comment (TimDiekmann) [inline packages/libs/error-stack/src/ser.rs:35]: Do we want to store `AttachmentKind` instead of `Frame` then?

File: packages/libs/error-stack/src/ser.rs [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+            }
+            FrameKind::Attachment(AttachmentKind::Printable(attachment)) => {
+                format!("{attachment}").serialize(serializer)
+            }
+        }
+    }
+}
+
+struct SerializeAttachments<'a, 'b>(&'a [&'b Frame]);
```
Comment (TimDiekmann) [inline packages/libs/error-stack/src/ser.rs:49]: I see why you named it this way. However, usually I find it hard to parse because it's easy to miss the "s". Do we want to name it `SerializeAttachmentList`?

File: packages/libs/error-stack/src/ser.rs [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+//! }
+//! ```
+
+use alloc::{format, vec, vec::Vec};
+
+use serde::{ser::SerializeMap, Serialize, Serializer};
+
+use crate::{AttachmentKind, Context, Frame, FrameKind, Report};
+
+struct SerializeAttachment<'a>(&'a Frame);
```
Comment (TimDiekmann) [inline packages/libs/error-stack/src/ser.rs:22]: Is there a specific reason we don't implement `Serialize` for `Frame`? I could see value in this being used directly. This could also be done later.

File: packages/libs/error-stack/src/ser.rs [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.collect_seq(
+            self.0
+                .iter()
+                .copied()
+                .filter(|attachment| {
+                    // for now opaque attachments are ignored
```
Comment (TimDiekmann) [inline packages/libs/error-stack/src/ser.rs:61]: Is this `unreachable!`?

File: packages/libs/error-stack/src/ser.rs [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+        #[allow(clippy::match_same_arms)]
+        match frame.kind() {
+            FrameKind::Context(_) => {
+                // `SerializeContext` ensures that no context is ever serialized
+                unreachable!()
+            }
+            FrameKind::Attachment(AttachmentKind::Opaque(_)) => {
+                // TODO: for now opaque attachments are unsupported, upcoming PR will fix that
+                // `SerializeContext` ensures that no such attachment is added
+                unreachable!()
```
Comment (TimDiekmann) [inline packages/libs/error-stack/src/ser.rs:40]: Maybe `todo`?
```suggestion
                todo!()
```

File: packages/libs/error-stack/tests/common.rs [hunk 10]
```diff
@@ -1,9 +1,5 @@
 #![allow(dead_code)]
 
-pub fn create_report() -> Report<RootError> {
-    Report::new(RootError)
-}
-
```
Comment (TimDiekmann) [inline packages/libs/error-stack/tests/common.rs:6]: Can we keep this at the top, so the output will not change at any test change?

**Review by TimDiekmann — CHANGES_REQUESTED** [review]
This looks very good, thank you, I'm super excited about this! Sorry, that it took me so long to review this 🙁 

Only have a few comments on the implementation details, the output LGTM 🎉

File: packages/libs/error-stack/src/ser.rs (hunk 7 above)
Comment (indietyp) [reply packages/libs/error-stack/src/ser.rs:22]: Yes, because we cannot guarantee that `Frame` serializes correctly. In that case, we would need a fallback (like `opaque`) type of frame.

File: packages/libs/error-stack/src/ser.rs (hunk 5 above)
Comment (indietyp) [reply packages/libs/error-stack/src/ser.rs:35]: This is in prep for hooks, where we will end up `request_ref` and `downcast_ref` (like with hooks).

File: packages/libs/error-stack/src/lib.rs (hunk 1 above)
Comment (indietyp) [reply packages/libs/error-stack/src/lib.rs]: fixed in [0147679](https://github.com/hashintel/hash/pull/1290/commits/014767941aa81eb87e56bdfc21492f85ee8eae06)

File: packages/libs/error-stack/src/ser.rs (hunk 8 above)
Comment (indietyp) [reply packages/libs/error-stack/src/ser.rs]: How so?

File: packages/libs/error-stack/src/ser.rs (hunk 5 above)
Comment (indietyp) [reply packages/libs/error-stack/src/ser.rs]: clarified in [0414ac3](https://github.com/hashintel/hash/pull/1290/commits/0414ac33a2e77fd9a0b6d8675329e9d52573cab4)

File: packages/libs/error-stack/src/ser.rs (hunk 6 above)
Comment (indietyp) [reply packages/libs/error-stack/src/ser.rs]: clarified in [43279ba](https://github.com/hashintel/hash/pull/1290/commits/43279baa1177e08e4826dcad20d6ac0cd31130b6)

File: packages/libs/error-stack/src/ser.rs (hunk 8 above)
Comment (indietyp) [reply packages/libs/error-stack/src/ser.rs]: clarified in [4c67564](https://github.com/hashintel/hash/pull/1290/commits/4c675645442c693dbd0877bef178c065b48bc2ff)

File: packages/libs/error-stack/tests/common.rs (hunk 10 above)
Comment (indietyp) [reply packages/libs/error-stack/tests/common.rs]: fixed in [4c67564](https://github.com/hashintel/hash/pull/1290/commits/4c675645442c693dbd0877bef178c065b48bc2ff)

File: packages/libs/error-stack/src/serde.rs [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+//!
+//! The value can be of any type, currently only printable attachments and context are supported, in
+//! the near future any values will be supported through the use of hooks.
+//!
+//! The serialized [`Report`] is a list of all current sources with the following output:
+//!
+//! ```json
+//! {
+//!     "context": "context display output",
+//!     "attachments": ["all attachments leading up to this context"],
```
Comment (TimDiekmann) [inline packages/libs/error-stack/src/serde.rs:11]: Nit, feel free to ignore
```suggestion
//!     "attachments": ["all", "attachments", "leading", "up", "to", "this", "context"],
```

**Review by TimDiekmann — APPROVED** [review]

**Comment by TimDiekmann** [comment]
Thanks @indietyp for implementing this feature, I'm super excited about that!

File: packages/libs/error-stack/src/ser.rs (hunk 8 above)
Comment (TimDiekmann) [reply packages/libs/error-stack/src/ser.rs]: Thank you 🙂

**Review by TimDiekmann — APPROVED** [review]
